Library/Homebrew/cmd/shellenv.sh

```diff
@@ -6,7 +6,7 @@
 #:  Consider adding evaluation of this command's output to your dotfiles (e.g. `~/.profile`, `~/.bash_profile`, or `~/.zprofile`) with: `eval $(brew shellenv)`
 
 homebrew-shellenv() {
-  case "$(/bin/ps -p $PPID -o comm=)" in
+  case "$(/bin/ps -p ${PPID:-$fish_pid} -c -o comm=)" in
```

Comment:
```suggestion
  case "$(/bin/ps -p $PPID -c -o comm=)" in
```
`homebrew-shellenv` is called in a _bash_ context, not fish, so this part of your analysis:
> The first problem this PR fixes is that, because fish doesn't set `PPID` and bash doesn't export it

is incorrect. `PPID` is _always_ set (by bash), and `fish_pid` is _never_ set (fish doesn't export it as an environment variable):
```
$ head test*
==> test1.fish <==
#!/usr/bin/env fish
/bin/ps -p $fish_pid -c
./test2.sh

==> test2.sh <==
#!/usr/bin/env bash
echo "fish_pid = $fish_pid"
/bin/ps -p $PPID -c

$ ./test1.fish
  PID TTY           TIME CMD
98534 ttys000    0:00.04 fish
fish_pid = 
  PID TTY           TIME CMD
98534 ttys000    0:00.04 fish
```

Reply:
And yet `brew shellenv` doesn't work properly for me with just `$PPID` there, and works as expected when I change it to `${PPID:-$fish_pid}`. Do you have an alternative solution?

Reply:
Just to be sure, what does `type -p brew` output in your fish shell?

Reply:
Also, what's the output of `brew config`?

Reply:
```
> type -p brew
/usr/local/bin/brew
> brew config
HOMEBREW_VERSION: 2.7.5
ORIGIN: https://github.com/Homebrew/brew
HEAD: bca4804a9e48de5319383d3eddadaa7f054c77da
Last commit: 3 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 9bb1a7cc9492e7aafc6c80f75810c3f5721b1578
Core tap last commit: 10 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_AUTO_UPDATE_SECS: 600
HOMEBREW_BAT: set
HOMEBREW_BAT_CONFIG_PATH: /Users/zozo/.config/bat/config_homebrew
HOMEBREW_BROWSER: open
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY_INSTALL_TIMES: set
HOMEBREW_EDITOR: vim
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_UPDATE_REPORT_ONLY_INSTALLED: set
HOMEBREW_UPDATE_TO_TAG: set
Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby
CPU: 16-core 64-bit dunno
Clang: 12.0 build 1200
Git: 2.30.0 => /usr/local/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 10.15.7-x86_64
CLT: 12.0.32.28
Xcode: 12.3
```

Reply:
OK, please add the following lines to the beginning of `homebrew-shellenv()` in `shellenv.sh`:
```
homebrew-shellenv() {
  # Check what variables are set
  if [[ -n $PPID ]]; then
    declare -p PPID fish_pid
  else
    # Something's odd, dump everything...
    declare -p
  fi
```
then run `brew shellenv 2>&1 | tee out.log`, then post the contents of `out.log`.

Here's what happens on my setup, without your PR's changes:
```
$ fish
Welcome to fish, the friendly interactive shell
Type `help` for instructions on how to use fish

> brew shellenv
declare -ir PPID="3192"
/usr/local/Homebrew/Library/Homebrew/cmd/shellenv.sh: line 10: declare: fish_pid: not found
set -gx HOMEBREW_PREFIX "/usr/local";
set -gx HOMEBREW_CELLAR "/usr/local/Cellar";
set -gx HOMEBREW_REPOSITORY "/usr/local/Homebrew";
set -q PATH; or set PATH ''; set -gx PATH "/usr/local/bin" "/usr/local/sbin" $PATH;
set -q MANPATH; or set MANPATH ''; set -gx MANPATH "/usr/local/share/man" $MANPATH;
set -q INFOPATH; or set INFOPATH ''; set -gx INFOPATH "/usr/local/share/info" $INFOPATH;
```

Reply:
`out.log`, without this PR:

```
declare -ir PPID="863"
/usr/local/Homebrew/Library/Homebrew/cmd/shellenv.sh: line 11: declare: fish_pid: not found
export HOMEBREW_PREFIX="/usr/local";
export HOMEBREW_CELLAR="/usr/local/Cellar";
export HOMEBREW_REPOSITORY="/usr/local/Homebrew";
export PATH="/usr/local/bin:/usr/local/sbin${PATH+:$PATH}";
export MANPATH="/usr/local/share/man${MANPATH+:$MANPATH}:";
export INFOPATH="/usr/local/share/info:${INFOPATH:-}";
```

And if I do the same *with* this PR in place:

```
declare -ir PPID="863"
/usr/local/Homebrew/Library/Homebrew/cmd/shellenv.sh: line 11: declare: fish_pid: not found
set -gx HOMEBREW_PREFIX "/usr/local";
set -gx HOMEBREW_CELLAR "/usr/local/Cellar";
set -gx HOMEBREW_REPOSITORY "/usr/local/Homebrew";
set -q PATH; or set PATH ''; set -gx PATH "/usr/local/bin" "/usr/local/sbin" $PATH;
set -q MANPATH; or set MANPATH ''; set -gx MANPATH "/usr/local/share/man" $MANPATH;
set -q INFOPATH; or set INFOPATH ''; set -gx INFOPATH "/usr/local/share/info" $INFOPATH;
```